src/js/protocols/stm32usbdfu.js

```diff
@@ -583,22 +593,30 @@ STM32DFU_protocol.prototype.upload_procedure = function (step) {
                             });
                         }
                     } else if (typeof chipInfo.external_flash !== "undefined") {
-                        // external flash, flash to the 3rd partition.
+                        // external flash, exst images must match the size of the partition being flashed.
+                        //
+                        // since the actual addresses to be flashed are in the .hex file this is not ideal
+                        // really we just need to make sure the flash chip contains addresses/offsets that match the ones in the .hex file.
+
                         self.chipInfo = chipInfo;
                         self.flash_layout = chipInfo.external_flash;
+                        self.available_flash_size = 0;
 
-                        var firmware_partition_index = 2;
-                        var firmware_sectors = self.flash_layout.sectors[firmware_partition_index];
-                        var firmware_partition_size = firmware_sectors.total_size;
+                        // Find a partition that matches the size of the hex.
+                        for (const firmwareSectors of self.flash_layout.sectors) {
 
-                        self.available_flash_size = firmware_partition_size;
+                            const firmwarePartitionSize = firmwareSectors.total_size;
+
+                            if (firmwarePartitionSize === self.hex.bytes_total) {
```

Comment:
What is the reason for matching this by size? This seems to be contrary to the way Intel Hex files are designed to work, and not really scalable / extensible:
- what happens if there are multiple partitions of the same size;
- what happens if the hex file is shorter than the partition it should go into;
- why is this not using the address, which is the way Intel Hex is meant to work.

Reply:
1. Currently first matching partition is used, by design.  There are no targets like this anyway.
2. Short hex results in no match, intended.  All current EXST HEX images are the same size as the partition.
3. I agree!  The pattern of checking the size come from the non-external flash code.  To be honest the whole size checking should probably just be deleted since the next step of the dfu flashing processes deals with the actual addresses.  If it was up to me I would ditch .hex files and use a zipped .json file and improve the entire release and flashing process and allow multiple images/memory areas to be flashed/updated.  similar to how PX4 does it.  I'll be happy to delete the size checking code from both the internal and external flash use cases and make it rely on addresses only, however this would take some time and I've got other outstanding work right now.  Suggest merging this stop-gap solution until time allows for further improvements.

Reply:
With regard to 3 above.  This section can be deleted (mostly):
https://github.com/betaflight/betaflight-configurator/blob/26ec8fb0e241c385f7c2a0c67c4e4ff1a86b673a/src/js/protocols/stm32usbdfu.js#L572-L629

This section could could be improved (out of scope of this PR):

https://github.com/betaflight/betaflight-configurator/blob/26ec8fb0e241c385f7c2a0c67c4e4ff1a86b673a/src/js/protocols/stm32usbdfu.js#L777-L804

As you can see, all the addresses in the .hex are checked to see if they match an 'erasable page'.  Instead of checking the size which I fully agree is wrong for both internal an external flash chips as the hex can contain non-contiguous memory addresses, the code could be updated to show 'address not found' or 'address not writable' error messages instead.